src/peergos/server/storage/FileContentAddressedStorage.java

```diff
@@ -0,0 +1,263 @@
+
+    @Override
+    public CompletableFuture<List<Multihash>> recursivePin(Multihash h) {
+        return CompletableFuture.completedFuture(Arrays.asList(h));
+    }
+
+    @Override
+    public CompletableFuture<List<Multihash>> recursiveUnpin(Multihash h) {
+        return CompletableFuture.completedFuture(Arrays.asList(h));
+    }
+
+    @Override
+    public CompletableFuture<List<Multihash>> getLinks(Multihash root) {
+        if (root instanceof Cid && ((Cid) root).codec == Cid.Codec.Raw)
+            return CompletableFuture.completedFuture(Collections.emptyList());
+        return get(root).thenApply(opt -> opt
+                .map(cbor -> cbor.links())
+                .orElse(Collections.emptyList())
+        );
+    }
+
+    private Path getFilePath(Multihash h) {
+        String name = h.toString();
+        int depth = 5;
+        Path path = Paths.get("");
+        for (int i=0; i < depth; i++)
+            path = path.resolve(Character.toString(name.charAt(i)));
+        // include full name in filename
+        path = path.resolve(name);
+        return path;
+    }
+
+    @Override
+    public CompletableFuture<Optional<CborObject>> get(Multihash hash) {
+        if (hash instanceof Cid && ((Cid) hash).codec == Cid.Codec.Raw)
+            throw new IllegalStateException("Need to call getRaw if cid is not cbor!");
+        return getRaw(hash).thenApply(opt -> opt.map(CborObject::fromByteArray));
+    }
+
+    @Override
+    public CompletableFuture<Optional<byte[]>> getRaw(Multihash hash) {
+        try {
+            Path path = getFilePath(hash);
+            File file = root.resolve(path).toFile();
+            if (! file.exists()){
+                return CompletableFuture.completedFuture(Optional.empty());
+            }
+            try (DataInputStream din = new DataInputStream(new BufferedInputStream(new FileInputStream(file)))) {
+                return CompletableFuture.completedFuture(Optional.of(Serialize.readFully(din)));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e.getMessage(), e);
+        }
+    }
+
+    public Multihash put(byte[] data, boolean isRaw) {
+        try {
+            Multihash rawhash = new Multihash(Multihash.Type.sha2_256, RAMStorage.hash(data));
+            Cid cid = new Cid(CID_V1, isRaw ? Cid.Codec.Raw : Cid.Codec.DagCbor, rawhash);
+            Path filePath = getFilePath(cid);
+            Path target = root.resolve(filePath);
+            Path parent = target.getParent();
+            File parentDir = parent.toFile();
+            if (! parentDir.exists() && ! parentDir.mkdirs())
+                throw new IllegalStateException("Couldn't create directory: " + parent);
+            for (Path someParent = parent; !someParent.equals(root); someParent = someParent.getParent()) {
+                final File someParentFile = someParent.toFile();
+                if ("root".equals(System.getProperty("user.name")) || ! someParentFile.canWrite()) {
```

Comment:
1. Why check if  the user is ``root``? 
2. This is not a robust way to do that.
3. Should probably add a note somewhere about this not  working on platforms other than linux.